lib/core/util.c

```diff
@@ -422,10 +422,26 @@ h2o_iovec_t h2o_build_destination(h2o_req_t *req, const char *prefix, size_t pre
     }
 
     /* append suffix path and query */
-    parts[num_parts++] = h2o_uri_escape(
-        &req->pool, req->path_normalized.base + req->pathconf->path.len, req->path_normalized.len - req->pathconf->path.len, "/@:");
-    if (req->query_at != SIZE_MAX)
-        parts[num_parts++] = h2o_iovec_init(req->path.base + req->query_at, req->path.len - req->query_at);
+
+    if (use_path_normalized) {
+        /* When proxying, we don't want to modify the input URL, unless the url
+           has been rewritten already */
+        parts[num_parts++] = h2o_uri_escape(&req->pool, req->path_normalized.base + req->pathconf->path.len,
+                                            req->path_normalized.len - req->pathconf->path.len, "/@:");
+        if (req->query_at != SIZE_MAX) {
+            parts[num_parts++] = h2o_iovec_init(req->path.base + req->query_at, req->path.len - req->query_at);
+        }
+    } else {
+        if (req->path.len > 1) {
+            size_t to_skip;
+            if (req->norm_indexes) {
+                to_skip = req->norm_indexes[req->pathconf->path.len - 1];
+            } else {
+                to_skip = req->pathconf->path.len;
+            }
+            parts[num_parts++] = (h2o_iovec_t){req->path.base + to_skip, req->path.len - to_skip};
```

Comment:
Please use `h2o_iovec_init` for constructing `h2o_iovec_t` to check that it is initialized correctly at compile time.

Compound initialization in C (i.e. `(struct foo){...}`) allows omission of following members, however in case of `h2o_iovec_t` we _must_ initialize all the properties.